packages/compass-connections/src/hooks/use-connection-status.ts

```diff
@@ -0,0 +1,44 @@
+import { useConnectionInfo } from '@mongodb-js/connection-storage/provider';
+import { useConnectionsManagerContext } from '../provider';
+import {
+  ConnectionStatus,
+  ConnectionsManagerEvents,
+} from '../connections-manager';
+
+import { type ConnectionInfo } from '@mongodb-js/connection-info';
+import { useEffect, useState } from 'react';
+
+export function useConnectionStatus(connectionInfo?: ConnectionInfo): {
```

Comment:
This interface seems too broad: it requires us to get the whole connection info before being able to access the status while the only required value here is an id, can we limit this to actually required parameters?

Reply:
Is there any reason to do so? I think it's more safe to just:

```
const connection = useConnectionInfo()
const status = useConnectionStatus(connection)
```

than

```
const connection = useConnectionInfo()
const status = useConnectionStatus(connection.id)
```

as requiring just the id is just an implementation detail of useConnectionStatus.

Reply:
Can you elaborate what does it mean for id to be an implementation detail in this case? How else other parts of the UI should be able to communicate that they are attached to a certain connection?

The reason to do so is that we don't want to make a requirement for some parts of the application to depend on the whole connection info object (that they generally don't care about), potentially causing unneccessary subscriptions to the state updates, heavy deep check to bypass referencial equality issues, and complex state propagation scenarios in cases where passing an id that uniquely identifies a connection should be enough.

Reply:
> The reason to do so is that we don't want to make a requirement for some parts of the application to depend on the whole connection info object (that they generally don't care about), potentially causing unneccessary subscriptions to the state updates, heavy deep check to bypass referencial equality issues, and complex state propagation scenarios in cases where passing an id that uniquely identifies a connection should be enough.

That's not the case, and using objects as an argument to a function is not a problem. If you have a hook that has a dependency on the whole object in the dependency array (thanks React for that) yes, it can be problematic, but there is no subscription in this case anywhere. However, by passing the whole object, you gain the benefit of:

1. Encapsulation, if the object changes or the hook depends on other fields of ConnectionInfo, you just change the hook.
2. Intention is clear: you get the status of a connection, that is what you want to achieve, and the code clearly makes that explicit (doesn't care about the id).
3. We already have a reference to an object when we use useConnectionInfo (the only way to get the connection info), and in JavaScript mostly everything (unless some optimisations) is passed by reference. In this case, there is no additional impact.

If I'm missing something obvious I can change the code obviously, but I don't think it will generate additional subscriptions to anything by just using an object as an argument instead of a string.

Reply:
> That's not the case, and using objects as an argument to a function is not a problem.

This is normal for functions, but you are not writing a function, you are writing a hook and hooks have a very spefic use-case: to be called inside React components.

> If you have a hook that has a dependency on the whole object in the dependency array (thanks React for that) yes, it can be problematic, but there is no subscription in this case anywhere

You are ignoring that to get that value to React rendering context to be able to pass it to the hook, it needs to pass through the props first (or get pulled from context), both cases mean that you are binding the component to state of a connection info

> Encapsulation, if the object changes or the hook depends on other fields of ConnectionInfo, you just change the hook

Realistically speaking, do we expect an id on the connecion info object not to be a sole identifier of a connection info? We should have a clear, documented way to identify a connection in the list of others. This is an id right now. Do we have any plans for connection info to be identified by something else? If yes, why would we design a system for ourselves that is poorly adapted to the other parts of the application code?

> We already have a reference to an object when we use useConnectionInfo (the only way to get the connection info), and in JavaScript mostly everything (unless some optimisations) is passed by reference. In this case, there is no additional impact.

Referential equality is currently preserved ONLY in the case of the old, single connection flow that this project will remove. Your multi connection code right now is written like this: connection info changes > code re-reads the list of connections from disk as the only existing way of updating the state > connection info object is completely overriden, you lost your reference (and due to the fact that the stores and hooks are all using local UI React state you have no referential equality between them). Same thing is implied by redux: changes in object require to create a new object reference, and state passed to the UI should be granular. As soon as we move this state to redux where it belongs, you will not have referential equality preserved even if we fix how the connectionInfo is updated. Relying on this is very brittle and will lead to hard-to-debug issues.

Reply:
Well the React magic on hooks arguments is something that I wasn't aware of and it's pretty unexpected. I appreciate the explanation because it's another design trap from the library.

I'll use the connectionInfoId as a parameter and resolve this comment.

Thanks a lot!

Reply:
It's not about the hook function arguments, there is no magic there. To get the value to the hook you need to get it inside the render method from somewhere first. For example in this scenario currently it's the `useConnectionInfo` hook that is "dumb" enough in the current single connection flow implementation that the value set in the application root never changes after connection. But this will not hold up for the multiple connections scenario, it's already not holding up, the code that's currently written doesn't give any referential equality guarantees.